File: electron/db/repositories/EpisodeRepository.ts
/**
 * EpisodeRepository：工作事件数据访问层
 * 含 userEdited 保护逻辑：若 user_edited=1 则 setOneLineSummary 拒绝覆盖并返回 false。
 */
import { randomUUID } from 'node:crypto'
import type { Episode, EntityRef } from '@/types'
import { getDatabase } from '../database'
import { parseJsonArray, stringifyJsonArray } from '../json'

interface EpisodeRow {
  id: string
  date: string
  start_time: string
  end_time: string
  title: string
  one_line_summary: string
  segment_ids: string
  entities: string
  topics: string
  user_edited: number
  report_eligible: number
  wiki_eligible: number
}

function rowToEpisode(row: EpisodeRow): Episode {
  return {
    id: row.id,
    date: row.date,
    startTime: row.start_time,
    endTime: row.end_time,
    title: row.title,
    oneLineSummary: row.one_line_summary,
    segmentIds: parseJsonArray<string>(row.segment_ids),
    entities: parseJsonArray<EntityRef>(row.entities),
    topics: parseJsonArray<string>(row.topics),
    userEdited: row.user_edited === 1,
    reportEligible: row.report_eligible === 1,
    wikiEligible: row.wiki_eligible === 1
  }
}

interface EpisodeParams {
  id: string
  date: string
  start_time: string
  end_time: string
  title: string
  one_line_summary: string
  segment_ids: string
  entities: string
  topics: string
  user_edited: number
  report_eligible: number
  wiki_eligible: number
}

function episodeToParams(episode: Episode): EpisodeParams {
  return {
    id: episode.id,
    date: episode.date,
    start_time: episode.startTime,
    end_time: episode.endTime,
    title: episode.title,
    one_line_summary: episode.oneLineSummary,
    segment_ids: stringifyJsonArray(episode.segmentIds),
    entities: stringifyJsonArray(episode.entities),
    topics: stringifyJsonArray(episode.topics),
    user_edited: episode.userEdited ? 1 : 0,
    report_eligible: episode.reportEligible ? 1 : 0,
    wiki_eligible: episode.wikiEligible ? 1 : 0
  }
}

export const EpisodeRepository = {
  insert(episode: Episode): Episode {
    const db = getDatabase()
    const id = episode.id || randomUUID()
    const params = episodeToParams({ ...episode, id })
    db.prepare(
      `INSERT INTO episodes (
        id, date, start_time, end_time, title, one_line_summary,
        segment_ids, entities, topics, user_edited, report_eligible, wiki_eligible
      ) VALUES (
        @id, @date, @start_time, @end_time, @title, @one_line_summary,
        @segment_ids, @entities, @topics, @user_edited, @report_eligible, @wiki_eligible
      )`
    ).run(params)
    const created = this.getById(id)
    if (!created) throw new Error(`Episode insert failed for id=${id}`)
    return created
  },

  update(id: string, patch: Partial<Episode>): Episode | null {
    const existing = this.getById(id)
    if (!existing) return null
    const merged: Episode = { ...existing, ...patch, id }
    const params = episodeToParams(merged)
    const db = getDatabase()
    db.prepare(
      `UPDATE episodes SET
        date = @date, start_time = @start_time, end_time = @end_time, title = @title,
        one_line_summary = @one_line_summary, segment_ids = @segment_ids, entities = @entities,
        topics = @topics, user_edited = @user_edited, report_eligible = @report_eligible,
        wiki_eligible = @wiki_eligible
      WHERE id = @id`
    ).run(params)
    return this.getById(id)
  },

  getById(id: string): Episode | null {
    const db = getDatabase()
    const row = db.prepare('SELECT * FROM episodes WHERE id = ?').get(id) as EpisodeRow | undefined
    return row ? rowToEpisode(row) : null
  },

  getByDate(date: string): Episode[] {
    const db = getDatabase()
    const rows = db
      .prepare('SELECT * FROM episodes WHERE date = ? ORDER BY start_time ASC')
      .all(date) as EpisodeRow[]
    return rows.map(rowToEpisode)
  },

  /** 按日期范围 [startDate, endDate] 查询 Episodes（含两端） */
  getByDateRange(startDate: string, endDate: string): Episode[] {
    const db = getDatabase()
    const rows = db
      .prepare(
        `SELECT * FROM episodes WHERE date >= ? AND date <= ? ORDER BY date ASC, start_time ASC`
      )
      .all(startDate, endDate) as EpisodeRow[]
    return rows.map(rowToEpisode)
  },

  /** 获取最近 N 天的 Episodes（含今日） */
  getRecent(days: number): Episode[] {
    const end = new Date()
    const start = new Date(end.getTime() - days * 24 * 60 * 60 * 1000)
    const fmt = (d: Date): string => {
      const y = d.getFullYear()
      const m = String(d.getMonth() + 1).padStart(2, '0')
      const day = String(d.getDate()).padStart(2, '0')
      return `${y}-${m}-${day}`
    }
    return this.getByDateRange(fmt(start), fmt(end))
  },

  /** 获取全库所有 Episodes（按日期升序） */
  getAll(): Episode[] {
    const db = getDatabase()
    const rows = db
      .prepare('SELECT * FROM episodes ORDER BY date ASC, start_time ASC')
      .all() as EpisodeRow[]
    return rows.map(rowToEpisode)
  },

  /**
   * 设置一句话总结。若 user_edited=1 则拒绝覆盖并返回 false（保护用户手动编辑）。
   * 当用户手动改写时，应通过 update(id, { oneLineSummary, userEdited: true }) 调用。
   */
  setOneLineSummary(id: string, summary: string): boolean {
    const db = getDatabase()
    const existing = this.getById(id)
    if (!existing) return false
    if (existing.userEdited) return false
    const result = db
      .prepare('UPDATE episodes SET one_line_summary = ? WHERE id = ? AND user_edited = 0')
      .run(summary, id)
    return result.changes > 0
  },

  setReportEligible(id: string, eligible: boolean): boolean {
    const db = getDatabase()
    const result = db
      .prepare('UPDATE episodes SET report_eligible = ? WHERE id = ?')
      .run(eligible ? 1 : 0, id)
    return result.changes > 0
  },

  setWikiEligible(id: string, eligible: boolean): boolean {
    const db = getDatabase()
    const result = db
      .prepare('UPDATE episodes SET wiki_eligible = ? WHERE id = ?')
      .run(eligible ? 1 : 0, id)
    return result.changes > 0
  },

  /**
   * 确认实体：标记 userConfirmed=true，使其不再被低置信过滤。
   * 按 type+name 匹配实体。返回更新后的 Episode，未找到返回 null。
   */
  confirmEntity(id: string, entityType: EntityRef['type'], entityName: string): Episode | null {
    const existing = this.getById(id)
    if (!existing) return null
    let found = false
    const entities = existing.entities.map((e) => {
      if (e.type === entityType && e.name === entityName) {
        found = true
        return { ...e, userConfirmed: true }
      }
      return e
    })
    if (!found) return null
    return this.update(id, { entities })
  },

  /**
   * 修正实体名：更新 name 并标记 userConfirmed=true。
   * 按 type+旧 name 匹配实体。返回更新后的 Episode，未找到返回 null。
   */
  correctEntity(
    id: string,
    entityType: EntityRef['type'],
    entityName: string,
    newName: string
  ): Episode | null {
    const existing = this.getById(id)
    if (!existing) return null
    let found = false
    const entities = existing.entities.map((e) => {
      if (e.type === entityType && e.name === entityName) {
        found = true
        return { ...e, name: newName, userConfirmed: true }
      }
      return e
    })
    if (!found) return null
    return this.update(id, { entities })
  },

  /**
   * 忽略实体：从 episode.entities 中移除匹配的实体。
   * 返回更新后的 Episode，未找到返回 null。
   */
  ignoreEntity(id: string, entityType: EntityRef['type'], entityName: string): Episode | null {
    const existing = this.getById(id)
    if (!existing) return null
    const before = existing.entities.length
    const entities = existing.entities.filter(
      (e) => !(e.type === entityType && e.name === entityName)
    )
    if (entities.length === before) return null
    return this.update(id, { entities })
  },

  hardDelete(id: string): boolean {
    const db = getDatabase()
    const result = db.prepare('DELETE FROM episodes WHERE id = ?').run(id)
    return result.changes > 0
  }
}

File: src-tauri/src/repositories/episode_repository.rs
//! EpisodeRepository：工作事件数据访问层（对应 electron/db/repositories/EpisodeRepository.ts）
//!
//! 含 userEdited 保护逻辑：若 user_edited=1 则 set_one_line_summary 拒绝覆盖并返回 false。

use rusqlite::{params, Connection};

use crate::db::database::get_database;
use crate::db::json::{parse_json_array, stringify_json_array};
use crate::models::{EntityRef, EntityRefType, Episode};

fn row_to_episode(row: &rusqlite::Row<'_>) -> rusqlite::Result<Episode> {
    let segment_ids_str: String = row.get("segment_ids")?;
    let entities_str: String = row.get("entities")?;
    let topics_str: String = row.get("topics")?;

    Ok(Episode {
        id: row.get("id")?,
        date: row.get("date")?,
        start_time: row.get("start_time")?,
        end_time: row.get("end_time")?,
        title: row.get("title")?,
        one_line_summary: row.get("one_line_summary")?,
        segment_ids: parse_json_array(&segment_ids_str),
        entities: parse_json_array(&entities_str),
        topics: parse_json_array(&topics_str),
        user_edited: row.get::<_, i64>("user_edited")? != 0,
        report_eligible: row.get::<_, i64>("report_eligible")? != 0,
        wiki_eligible: row.get::<_, i64>("wiki_eligible")? != 0,
        dominant_activity_type: None,
    })
}

fn insert_episode(conn: &Connection, episode: &Episode) -> anyhow::Result<()> {
    let segment_ids = stringify_json_array(&episode.segment_ids);
    let entities = stringify_json_array(&episode.entities);
    let topics = stringify_json_array(&episode.topics);

    conn.execute(
        "INSERT INTO episodes (
            id, date, start_time, end_time, title, one_line_summary,
            segment_ids, entities, topics, user_edited, report_eligible, wiki_eligible
        ) VALUES (
            ?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12
        )",
        params![
            episode.id,
            episode.date,
            episode.start_time,
            episode.end_time,
            episode.title,
            episode.one_line_summary,
            segment_ids,
            entities,
            topics,
            episode.user_edited as i64,
            episode.report_eligible as i64,
            episode.wiki_eligible as i64,
        ],
    )?;
    Ok(())
}

fn update_episode_full(conn: &Connection, episode: &Episode) -> anyhow::Result<()> {
    let segment_ids = stringify_json_array(&episode.segment_ids);
    let entities = stringify_json_array(&episode.entities);
    let topics = stringify_json_array(&episode.topics);

    conn.execute(
        "UPDATE episodes SET
            date = ?2, start_time = ?3, end_time = ?4, title = ?5,
            one_line_summary = ?6, segment_ids = ?7, entities = ?8,
            topics = ?9, user_edited = ?10, report_eligible = ?11,
            wiki_eligible = ?12
        WHERE id = ?1",
        params![
            episode.id,
            episode.date,
            episode.start_time,
            episode.end_time,
            episode.title,
            episode.one_line_summary,
            segment_ids,
            entities,
            topics,
            episode.user_edited as i64,
            episode.report_eligible as i64,
            episode.wiki_eligible as i64,
        ],
    )?;
    Ok(())
}

/// 仅更新 entities 字段
fn update_entities(conn: &Connection, id: &str, entities: &[EntityRef]) -> anyhow::Result<()> {
    let entities_str = stringify_json_array(entities);
    conn.execute(
        "UPDATE episodes SET entities = ?1 WHERE id = ?2",
        params![entities_str, id],
    )?;
    Ok(())
}

pub struct EpisodeRepository;

impl EpisodeRepository {
    pub fn insert(episode: Episode) -> anyhow::Result<Episode> {
        let id = if episode.id.is_empty() {
            uuid::Uuid::new_v4().to_string()
        } else {
            episode.id.clone()
        };
        let mut episode = episode;
        episode.id = id.clone();

        {
            let conn = get_database()?;
            insert_episode(&conn, &episode)?;
        }
        Self::get_by_id(&id)?
            .ok_or_else(|| anyhow::anyhow!("Episode insert failed for id={}", id))
    }

    pub fn update(id: &str, patch: Episode) -> anyhow::Result<Option<Episode>> {
        let existing = match Self::get_by_id(id)? {
            Some(e) => e,
            None => return Ok(None),
        };
        let merged = merge_episode(existing, patch, id);
        {
            let conn = get_database()?;
            update_episode_full(&conn, &merged)?;
        }
        Self::get_by_id(id)
    }

    pub fn get_by_id(id: &str) -> anyhow::Result<Option<Episode>> {
        let conn = get_database()?;
        let mut stmt = conn.prepare("SELECT * FROM episodes WHERE id = ?1")?;
        let mut rows = stmt.query(params![id])?;
        match rows.next()? {
            Some(row) => Ok(Some(row_to_episode(row)?)),
            None => Ok(None),
        }
    }

    pub fn get_by_date(date: &str) -> anyhow::Result<Vec<Episode>> {
        let conn = get_database()?;
        let mut stmt =
            conn.prepare("SELECT * FROM episodes WHERE date = ?1 ORDER BY start_time ASC")?;
        let episodes = stmt
            .query_map(params![date], row_to_episode)?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        Ok(episodes)
    }

    /// 按日期范围 [startDate, endDate] 查询 Episodes（含两端）
    pub fn get_by_date_range(start_date: &str, end_date: &str) -> anyhow::Result<Vec<Episode>> {
        let conn = get_database()?;
        let mut stmt = conn.prepare(
            "SELECT * FROM episodes WHERE date >= ?1 AND date <= ?2 ORDER BY date ASC, start_time ASC",
        )?;
        let episodes = stmt
            .query_map(params![start_date, end_date], row_to_episode)?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        Ok(episodes)
    }

    /// 获取最近 N 天的 Episodes（含今日）
    pub fn get_recent(days: i64) -> anyhow::Result<Vec<Episode>> {
        let end = chrono::Utc::now();
        let start = end - chrono::Duration::days(days);
        let fmt = |d: chrono::DateTime<chrono::Utc>| d.format("%Y-%m-%d").to_string();
        Self::get_by_date_range(&fmt(start), &fmt(end))
    }

    /// 获取全库所有 Episodes（按日期升序）
    pub fn get_all() -> anyhow::Result<Vec<Episode>> {
        let conn = get_database()?;
        let mut stmt =
            conn.prepare("SELECT * FROM episodes ORDER BY date ASC, start_time ASC")?;
        let episodes = stmt
            .query_map([], row_to_episode)?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        Ok(episodes)
    }

    /// 设置一句话总结。若 user_edited=1 则拒绝覆盖并返回 false。
    pub fn set_one_line_summary(id: &str, summary: &str) -> anyhow::Result<bool> {
        let existing = match Self::get_by_id(id)? {
            Some(e) => e,
            None => return Ok(false),
        };
        if existing.user_edited {
            return Ok(false);
        }
        let conn = get_database()?;
        let changes = conn.execute(
            "UPDATE episodes SET one_line_summary = ?1 WHERE id = ?2 AND user_edited = 0",
            params![summary, id],
        )?;
        Ok(changes > 0)
    }

    pub fn set_report_eligible(id: &str, eligible: bool) -> anyhow::Result<bool> {
        let conn = get_database()?;
        let changes = conn.execute(
            "UPDATE episodes SET report_eligible = ?1 WHERE id = ?2",
            params![eligible as i64, id],
        )?;
        Ok(changes > 0)
    }

    pub fn set_wiki_eligible(id: &str, eligible: bool) -> anyhow::Result<bool> {
        let conn = get_database()?;
        let changes = conn.execute(
            "UPDATE episodes SET wiki_eligible = ?1 WHERE id = ?2",
            params![eligible as i64, id],
        )?;
        Ok(changes > 0)
    }

    /// 确认实体：标记 user_confirmed=true
    pub fn confirm_entity(
        id: &str,
        entity_type: EntityRefType,
        entity_name: &str,
    ) -> anyhow::Result<Option<Episode>> {
        let mut existing = match Self::get_by_id(id)? {
            Some(e) => e,
            None => return Ok(None),
        };
        let mut found = false;
        for e in existing.entities.iter_mut() {
            if e.ref_type == entity_type && e.name == entity_name {
                e.user_confirmed = true;
                found = true;
            }
        }
        if !found {
            return Ok(None);
        }
        {
            let conn = get_database()?;
            update_entities(&conn, id, &existing.entities)?;
        }
        Self::get_by_id(id)
    }

    /// 修正实体名：更新 name 并标记 user_confirmed=true
    pub fn correct_entity(
        id: &str,
        entity_type: EntityRefType,
        entity_name: &str,
        new_name: &str,
    ) -> anyhow::Result<Option<Episode>> {
        let mut existing = match Self::get_by_id(id)? {
            Some(e) => e,
            None => return Ok(None),
        };
        let mut found = false;
        for e in existing.entities.iter_mut() {
            if e.ref_type == entity_type && e.name == entity_name {
                e.name = new_name.to_string();
                e.user_confirmed = true;
                found = true;
            }
        }
        if !found {
            return Ok(None);
        }
        {
            let conn = get_database()?;
            update_entities(&conn, id, &existing.entities)?;
        }
        Self::get_by_id(id)
    }

    /// 忽略实体：从 episode.entities 中移除匹配的实体
    pub fn ignore_entity(
        id: &str,
        entity_type: EntityRefType,
        entity_name: &str,
    ) -> anyhow::Result<Option<Episode>> {
        let existing = match Self::get_by_id(id)? {
            Some(e) => e,
            None => return Ok(None),
        };
        let before_len = existing.entities.len();
        let entities: Vec<EntityRef> = existing
            .entities
            .iter()
            .filter(|e| !(e.ref_type == entity_type && e.name == entity_name))
            .cloned()
            .collect();
        if entities.len() == before_len {
            return Ok(None);
        }
        {
            let conn = get_database()?;
            update_entities(&conn, id, &entities)?;
        }
        Self::get_by_id(id)
    }

    pub fn hard_delete(id: &str) -> anyhow::Result<bool> {
        let conn = get_database()?;
        let changes = conn.execute("DELETE FROM episodes WHERE id = ?1", params![id])?;
        Ok(changes > 0)
    }
}

/// 合并现有 episode 与 patch（patch 的非空字段覆盖 existing）
fn merge_episode(mut existing: Episode, patch: Episode, id: &str) -> Episode {
    if !patch.date.is_empty() {
        existing.date = patch.date;
    }
    if !patch.start_time.is_empty() {
        existing.start_time = patch.start_time;
    }
    if !patch.end_time.is_empty() {
        existing.end_time = patch.end_time;
    }
    if !patch.title.is_empty() {
        existing.title = patch.title;
    }
    if !patch.one_line_summary.is_empty() {
        existing.one_line_summary = patch.one_line_summary;
    }
    if !patch.segment_ids.is_empty() {
        existing.segment_ids = patch.segment_ids;
    }
    if !patch.entities.is_empty() {
        existing.entities = patch.entities;
    }
    if !patch.topics.is_empty() {
        existing.topics = patch.topics;
    }
    // 布尔字段总是采用 patch 值
    existing.user_edited = patch.user_edited;
    existing.report_eligible = patch.report_eligible;
    existing.wiki_eligible = patch.wiki_eligible;
    if patch.dominant_activity_type.is_some() {
        existing.dominant_activity_type = patch.dominant_activity_type;
    }
    existing.id = id.to_string();
    existing
}
